---
title: "Tomato_SimpleTidy_vs_WGCNA"
author: "Chenxin Li"
date: '2023-01-13'
output: html_notebook 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This work sheet aims to compare Li's SimpleTidy GeneCoEx workflow with the WGCNA package. 
[WGCNA](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559) is regarded as the "method of choice" by some for gene co-expression analyses. 
We will be benchmarking the two workflows using the tomato fruit developmental series data. 
Ref: [Shinozaki et al., 2018](https://www.nature.com/articles/s41467-017-02782-9).

# Dependencies
```{r}
library(tidyverse)
library(WGCNA)

library(ggalluvial)
library(ggbeeswarm)

library(patchwork)
library(RColorBrewer)
library(viridis)

set.seed(666)
```

I kind of ran into a dependency hell for WGCNA. I sorted it out using these steps:

1. Install BioConductor Manager: `install.packages("BiocManager")`
2. Install WGCNA from BioConductor: `BiocManager::install("WGCNA", force = T)`

Hopefully you don't run into this issue. If you do, hopefully my fix works for you. 

# Load data 
I already performed gene co-expression and clustering analyses using my "Simply Tidy Gene CoEx" workflow. 
I will just import the results from my analyses first. 

## Simple Tidy results 
```{r}
modules_mean_z_reorded <- read_csv("../Results/modules_mean_z_reordered.csv", col_types = cols())
my_network_modules <- read_csv("../Results/my_network_modules.csv", col_types = cols())
Exp_table_long_averaged_z_high_var_modules <- read_csv("../Results/Exp_table_long_averaged_z_high_var_modules.csv", col_types = cols())
```

I am reading in 3 tables.

1. `modules_mean_z_reordered` contains the mean z score of all genes in each module across developmental stages and tissues. 
2. `my_network_modules` contains the module membership for each gene in the analyses.
3. `Exp_table_long_averaged_z_high_var_modules` contains the z score of genes in each module. 

For WGCNA, I need a normalized gene expression matrix. Let's make that from the tpm table. 
```{r}
Exp_table <- read_csv("../Data/Shinozaki_tpm_representative_transcripts.csv", col_types = cols())
head(Exp_table)
```

For the rest of this workflow, I will be following this [tutorial for WCGNA](https://bioinformaticsworkbook.org/tutorials/wgcna.html#gsc.tab=0). 
 
# Gene selection for WGCNA
While it is not covered in the tutorial for WGCNA, Shinozaki et al. performed gene selection steps based on coefficient of variation (CV). 
CV is the ratio of sd over mean.  

> Co-expression network modules were identified using averaged RPM values and the WGCNA package (v1.51) in. 
Genes with a low coefficient of variation of averaged RPM (CV < 1) among all sample types (cell/tissue types, different latitudinal sections, or developmental stages) were discarded and the remaining 12,662 genes were used for the analysis.

We will do things slightly different here. Since we have TPM table, we will use averaged TPM values instead of RPM. 
Secondly, for the purpose of benchmarking, I will only use samples that are hand dissected (without using laser micro-dissection samples).
Thirdly, again for the purpose of benchmarking, I will only the top 5000 variable genes. 

Let's pull those top variable genes.  
## Wide to long 
```{r}
Exp_table_long <- Exp_table %>% 
  rename(gene_ID = `...1`) %>% 
  pivot_longer(cols = !gene_ID, names_to = "library", values_to = "tpm") %>% 
  mutate(logTPM = log10(tpm + 1)) 

head(Exp_table_long)
```

## Average up reps
To average up reps, we need the metadata. 
Let's load that. 
```{r}
Metadata <- read_excel("../Data/Shinozaki_datasets_SRA_info.xlsx")
head(Metadata)
```

```{r}
Exp_table_long_averaged <- Exp_table_long %>% 
  full_join(Metadata %>% 
              select(Run, `Sample Name`, tissue, dev_stage), 
            by = c("library"="Run")) %>% 
   mutate(dissection_method = case_when(
    str_detect(tissue, "epidermis") ~ "LM",
    str_detect(tissue, "Collenchyma") ~ "LM",
    str_detect(tissue, "Parenchyma") ~ "LM",
    str_detect(tissue, "Vascular") ~ "LM",
    str_detect(dev_stage, "Anthesis") ~ "LM",
    str_detect(dev_stage, "5 DPA") &
      str_detect(tissue, "Locular tissue|Placenta|Seeds") ~ "LM",
    T ~ "Hand"
  )) %>% 
  filter(dissection_method == "Hand") %>% 
  group_by(gene_ID, `Sample Name`, tissue, dev_stage) %>% 
  summarise(mean.logTPM = mean(logTPM)) %>% 
  ungroup()  

head(Exp_table_long_averaged)
```
As in the Simple Tidy workflow, we only took hand dissected samples. This step could take a while. 

## z score
```{r}
Exp_table_long_averaged_z <- Exp_table_long_averaged %>% 
  group_by(gene_ID) %>% 
  mutate(z.score = (mean.logTPM - mean(mean.logTPM))/sd(mean.logTPM)) %>% 
  ungroup()

head(Exp_table_long_averaged_z)
```


## Find top 5000 variable 
```{r}
high_var_genes <- Exp_table_long_averaged_z %>% 
  group_by(gene_ID) %>% 
  summarise(var = var(mean.logTPM)) %>% 
  ungroup() %>% 
  filter(var > quantile(var, 0.667))

high_var_genes5000 <- high_var_genes %>% 
  slice_max(order_by = var, n = 5000) 

head(high_var_genes5000)
dim(high_var_genes5000)
```

```{r}
Exp_table_long_averaged_z_high_var <- Exp_table_long_averaged_z %>% 
  filter(gene_ID %in% high_var_genes5000$gene_ID)

head(Exp_table_long_averaged_z_high_var)

Exp_table_long_averaged_z_high_var %>% 
  group_by(gene_ID) %>% 
  count() %>% 
  nrow()
```


# Input requirement for WGCNA 
WGCNA requires matrix with genes are columns and sample as names. Let's make that. 
```{r}
WGCNA_input_mat <- Exp_table_long_averaged_z_high_var %>% 
  mutate(treatment = paste(tissue, dev_stage, sep = ".")) %>% 
  select(gene_ID, treatment, mean.logTPM) %>% 
  pivot_wider(names_from = gene_ID, values_from = mean.logTPM) %>% 
  as.data.frame()

row.names(WGCNA_input_mat) <- WGCNA_input_mat$treatment
head(WGCNA_input_mat)
```

```{r}
input_mat <- WGCNA_input_mat[, -1] %>% as.matrix()
input_mat[1:6, 1:10]
```
# Diagnostic statistics
WGCNA has its edge selection method. It picks a threshold value, below which edges are removed. 
The math behind it is beyond me. But let's just follow the instructions. 
```{r}
allowWGCNAThreads(4)
```

```{r}
powers <- c(c(1:20), seq(from = 1, to = 20, by = 2))

soft <- pickSoftThreshold(
  input_mat,             
  powerVector = powers,
  verbose = 5
  )
```

```{r}
WGCNA_power1 <- soft$fitIndices %>% 
  ggplot(aes(x = Power, y = SFT.R.sq )) +
  geom_line(color = "grey20", size = 1.1, alpha = 0.8) +
  geom_point(size = 3, alpha = 0.8, color = "dodgerblue") +
  labs(x = "Soft Threshold (power)",
       y = "Scale Free Topology\nModel Fit signed R^2",
       title = "Data from Shinozaki et al., 2018") +
  theme_classic() +
  theme(
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    plot.title = element_text(size = 12)
  )


WGCNA_power2 <- soft$fitIndices %>% 
  ggplot(aes(x = Power, y = mean.k. )) +
  geom_line(color = "grey20", size = 1.1, alpha = 0.8) +
  geom_point(size = 3, alpha = 0.8, color = "violetred2") +
  labs(x = "Soft Threshold (power)",
       y = "Mean Connectivity") +
  theme_classic() +
  theme(
    text = element_text(size = 14),
    axis.text = element_text(color = "black")
  )

wrap_plots(WGCNA_power1, WGCNA_power2,
           ncol = 2)

ggsave("../Results/WGCNA_tomato_power.svg", width = 6, height = 3, bg = "white")
ggsave("../Results/WGCNA_tomato_power.png", width = 6, height = 3, bg = "white")
```
Let's go with power of 15, where the curves really flattern out.  

# Detect modules 
```{r}
picked_power <- 15
temp_cor <- cor
cor <- WGCNA::cor

tomato_WGCNA_network <- blockwiseModules(
  input_mat,
  
  power = picked_power, 
  TOMType = "signed", 
  networkType = "signed",
  
  pamRespectsDendro = F,
  minModuleSize = 30,
  
  mergeCutHeight = 0.25,
  
)  

cor <- temp_cor
```


# Module membership
```{r}
module_df <- data.frame(
  gene.id = names(tomato_WGCNA_network$colors),
  module = tomato_WGCNA_network$colors
)

head(module_df)
module_df %>% 
  group_by(module) %>% 
  count()
```
We got 14 modules. 

# Module-treatment correspondance 
```{r}
Module_eigen <- moduleEigengenes(input_mat, tomato_WGCNA_network$colors)$eigengenes
```

## Heat map representation 
### Reorder rows and columns 
```{r}
Module_eigen_long <- Module_eigen %>% 
  mutate(tag = row.names(.)) %>% 
  pivot_longer(cols = !tag, names_to = "Module", values_to = "Eigen") %>% 
  separate(tag, c("tissue", "dev_stage"), sep = "\\.") %>% 
  mutate(order_x = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>%  
  mutate(stage = case_when(
    str_detect(dev_stage, "MG|Br|Pk") ~ str_sub(dev_stage, start = 1, end = 2),
    T ~ dev_stage
  )) %>% 
  mutate(stage = factor(stage, levels = c(
   "5 DPA",
   "10 DPA",
   "20 DPA",
   "30 DPA",
   "MG",
   "Br",
   "Pk",
   "LR",
   "RR"
  ))) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x)) 


head(Module_eigen_long)
```

#### Module peak 
```{r}
Module_eigen_long_peak <- Module_eigen_long %>% 
  group_by(Module) %>% 
  slice_max(order_by = Eigen, n = 1) %>% 
  ungroup() %>% 
  mutate(order_y = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(Module = reorder(Module, -order_y)) %>% 
  mutate(peak_stage = dev_stage) %>% 
  select(Module, order_y, peak_stage)
   

head(Module_eigen_long_peak)
```
```{r}
Module_eigen_long_reordered <- Module_eigen_long %>% 
  full_join(Module_eigen_long_peak, by = "Module") %>% 
  mutate(colors = str_remove(Module, "ME")) %>% 
  mutate(Module = reorder(Module, -order_y)) %>% 
   mutate(colors = reorder(colors, -order_y)) 
```

### Outliers
```{r}
summary(Module_eigen_long_reordered$Eigen)
quantile(Module_eigen_long_reordered$Eigen, c(0.75,0.9, 0.95, 0.975, 0.99))
```

```{r}
tomato_WGCNA_heatmap <- Module_eigen_long_reordered %>% 
  mutate(Eigen_clipped = case_when(
    Eigen > 0.25 ~ 0.25,
    Eigen < -0.25 ~ -0.25,
    T ~ Eigen
  )) %>% 
  ggplot(aes(x = tissue, y = colors)) +
  facet_grid(.~ dev_stage, scales = "free", space = "free") +
  geom_tile(aes(fill = Eigen_clipped)) +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdBu")),
                       limits = c(-0.25, 0.25),
                       breaks = c(-0.25, 0, 0.25),
                       labels = c("<0.25", "0" , ">0.25")) +
  labs(x = NULL,
       y = "Module",
       fill = "Eigen Gene Exp.") +
  theme_classic() +
  theme(
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    axis.text.x = element_blank(),
    strip.text = element_blank(),
    legend.position = "top",
    panel.spacing = unit(0.5, "lines") 
  )

tomato_WGCNA_heatmap
```

### Make color strip for heatmap 
```{r}
module_df_colors <- Module_eigen_long_peak %>% 
  mutate(colors = str_remove(Module, "ME"))

module_df_colors

heat_map_color_strip_y <- module_df_colors %>% 
  ggplot(aes(x = "", y = Module)) +
  geom_tile(aes(fill = colors)) +
  scale_fill_identity() +
  theme_void()
```

```{r}
heatmap_color_strip1 <- expand.grid(
  tissue = unique(Module_eigen_long$tissue),
  dev_stage = unique(Module_eigen_long$dev_stage), 
  stringsAsFactors = F
) %>% 
  mutate(dev_stage = as.character(dev_stage)) %>% 
  filter(dev_stage != "Anthesis") %>% 
  filter(str_detect(tissue, "epider|chyma|Vasc") == F) %>% 
  filter((dev_stage == "5 DPA" &
           str_detect(tissue, "Locular tissue|Placenta|Seeds"))==F) %>% 
  filter((str_detect(dev_stage, "styla") &
           str_detect(tissue, "Colum"))==F) %>% 
  mutate(order_x = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(stage = case_when(
    str_detect(dev_stage, "MG|Br|Pk") ~ str_sub(dev_stage, start = 1, end = 2),
    T ~ dev_stage
  )) %>% 
  mutate(stage = factor(stage, levels = c(
   "5 DPA",
   "10 DPA",
   "20 DPA",
   "30 DPA",
   "MG",
   "Br",
   "Pk",
   "LR",
   "RR"
  ))) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x)) %>% 
  ggplot(aes(x = tissue, y = 1)) +
  facet_grid(.~ dev_stage, scales = "free", space = "free") +
  geom_tile(aes(fill = tissue)) +
  scale_fill_manual(values = brewer.pal(8, "Set2")) +
  guides(fill = guide_legend(nrow = 1)) +
  theme_void() +
  theme(
    legend.position = "bottom",
    strip.text = element_blank(),
    text = element_text(size = 14),
    panel.spacing = unit(0.5, "lines"),
    legend.key.height = unit(0.75, "lines")
  )


heatmap_color_strip2 <- expand.grid(
  tissue = unique(Module_eigen_long$tissue),
  dev_stage = unique(Module_eigen_long$dev_stage), 
  stringsAsFactors = F
) %>% 
  mutate(dev_stage = as.character(dev_stage)) %>% 
  filter(dev_stage != "Anthesis") %>% 
  filter(str_detect(tissue, "epider|chyma|Vasc") == F) %>% 
  filter((dev_stage == "5 DPA" &
           str_detect(tissue, "Locular tissue|Placenta|Seeds"))==F) %>% 
  filter((str_detect(dev_stage, "styla") &
           str_detect(tissue, "Colum"))==F) %>% 
  mutate(order_x = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(stage = case_when(
    str_detect(dev_stage, "MG|Br|Pk") ~ str_sub(dev_stage, start = 1, end = 2),
    T ~ dev_stage
  )) %>% 
  mutate(stage = factor(stage, levels = c(
   "5 DPA",
   "10 DPA",
   "20 DPA",
   "30 DPA",
   "MG",
   "Br",
   "Pk",
   "LR",
   "RR"
  ))) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x)) %>% 
  ggplot(aes(x = tissue, y = 1)) +
  facet_grid(.~ dev_stage, scales = "free", space = "free") +
  geom_tile(aes(fill = stage)) +
  scale_fill_manual(values = viridis(9, option = "D")) +
  labs(fill = "stage") +
  guides(fill = guide_legend(nrow = 1)) +
  theme_void() +
  theme(
    legend.position = "bottom",
    strip.text = element_blank(),
    text = element_text(size = 14),
    panel.spacing = unit(0.5, "lines"),
    legend.key.height = unit(0.75, "lines")
  )

blank <- module_df_colors %>% 
  ggplot(aes(x = "", y = Module)) +
  theme_void()

```

```{r}
wrap_plots(tomato_WGCNA_heatmap, heat_map_color_strip_y,
           heatmap_color_strip1, blank,
           heatmap_color_strip2, blank, 
           nrow = 3, ncol = 2, 
           heights = c(1, 0.05, 0.05),
           widths = c(1, 0.02), guides = "collect") &
  theme(
    legend.position = "bottom",
    legend.box = "vertical"
  )

ggsave("../Results/WGCNA_tomato_heatmap.svg", height = 4.8, width = 10, bg = "white")
ggsave("../Results/WGCNA_tomato_heatmap.png", height = 4.8, width = 10, bg = "white")
```
# Module QC
## Bait genes
We have two bait genes:

1. PG: Solly.M82.10G020850.1, involved in making the fruit softer
2. PSY1: Solly.M82.03G005440.5, involved in making the fruit red. 

Let's check which module(s) they are assigned to.  
```{r}
module_df %>% 
  filter(gene.id == "Solly.M82.10G020850.1" | 
           gene.id == "Solly.M82.03G005440.5")
```
They are both in module "blue". This is good to see, because we know they are involved in the same biological processes. 
I have no idea which one plum1 is looking at the heat map. 

## Line graphs
```{r}
Exp_table_long_averaged_modules <- Exp_table_long_averaged %>% 
  inner_join(module_df, by = c("gene_ID" = "gene.id")) %>% 
  inner_join(module_df_colors %>% 
               select(Module, colors), by = c("module"="colors")) %>% 
  group_by(gene_ID) %>% 
  mutate(z.score = (mean.logTPM - mean(mean.logTPM))/sd(mean.logTPM)) %>% 
  ungroup()

head(Exp_table_long_averaged_modules)
```

I computed z score here. Now I can graph a couple modules to check. 
We will do Module blue because that where our baits are. 
Let's pick another one. 
I am picking module turquoise because according to peak expression, it peaks at the earliest stage of development. 

Before we make the line graph, let's also make an average of z scores for all the modules. 
```{r}
tomato_WGCNA_modules_mean_z <- Exp_table_long_averaged_modules %>% 
  group_by(dev_stage, tissue, module) %>% 
  summarise(mean.z = mean(z.score)) %>% 
  ungroup() %>% 
  filter(is.na(mean.z)==F) %>% 
  mutate(order_x = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x))

head(tomato_WGCNA_modules_mean_z)
```


```{r}
module_line_plot <- Exp_table_long_averaged_modules %>% 
  mutate(order_x = case_when(
    str_detect(dev_stage, "5") ~ 1,
    str_detect(dev_stage, "10") ~ 2,
    str_detect(dev_stage, "20") ~ 3,
    str_detect(dev_stage, "30") ~ 4,
    str_detect(dev_stage, "MG") ~ 5,
    str_detect(dev_stage, "Br") ~ 6,
    str_detect(dev_stage, "Pk") ~ 7,
    str_detect(dev_stage, "LR") ~ 8,
    str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x)) %>% 
  filter(module == "blue" |
          module == "turquoise") %>% 
  ggplot(aes(x = dev_stage, y = z.score)) +
  facet_grid(module ~ tissue) +
  geom_line(aes(group = gene_ID), alpha = 0.3, color = "grey70") +
  geom_line(
    data = tomato_WGCNA_modules_mean_z %>% 
      filter(
        module == "blue" |
          module == "turquoise"
      ), 
    aes(y = mean.z, group = module),
    size = 1.1, alpha = 0.9
  ) +
  labs(x = NULL,
       y = "z score") +
  theme_classic() +
  theme(
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    axis.text.x = element_blank(),
    panel.spacing = unit(1, "line")
  )


module_lines_color_strip <- expand.grid(
  tissue = unique(Module_eigen_long$tissue),
  dev_stage = unique(Module_eigen_long$dev_stage), 
  stringsAsFactors = F
) %>% 
  mutate(dev_stage = as.character(dev_stage)) %>% 
  filter(dev_stage != "Anthesis") %>% 
  filter(str_detect(tissue, "epider|chyma|Vasc") == F) %>% 
  mutate(order_x = case_when(
        str_detect(dev_stage, "5") ~ 1,
        str_detect(dev_stage, "10") ~ 2,
        str_detect(dev_stage, "20") ~ 3,
        str_detect(dev_stage, "30") ~ 4,
        str_detect(dev_stage, "MG") ~ 5,
        str_detect(dev_stage, "Br") ~ 6,
        str_detect(dev_stage, "Pk") ~ 7,
        str_detect(dev_stage, "LR") ~ 8,
        str_detect(dev_stage, "RR") ~ 9
  )) %>% 
  mutate(stage = case_when(
    str_detect(dev_stage, "MG|Br|Pk") ~ str_sub(dev_stage, start = 1, end = 2),
    T ~ dev_stage
  )) %>% 
  mutate(stage = factor(stage, levels = c(
   "5 DPA",
   "10 DPA",
   "20 DPA",
   "30 DPA",
   "MG",
   "Br",
   "Pk",
   "LR",
   "RR"
  ))) %>% 
  mutate(dev_stage = reorder(dev_stage, order_x)) %>% 
  ggplot(aes(x = dev_stage, y = 1)) +
  facet_grid(. ~ tissue) +
  geom_tile(aes(fill = stage)) +
  scale_fill_manual(values = viridis(9, option = "D")) +
  theme_void() +
  theme(
    legend.position = "bottom",
    strip.text = element_blank(),
    text = element_text(size = 14),
    panel.spacing = unit(1, "lines")
  )


wrap_plots(module_line_plot, module_lines_color_strip,
           nrow = 2, heights = c(1, 0.08))

ggsave("../Results/WGCNA_tomato_module_line_plots.svg", height = 4, width = 8.2, bg = "white")
ggsave("../Results/WGCNA_tomato_module_line_plots.png", height = 4, width = 8.2, bg = "white")
```

# Heatmap for module z score

```{r}
tomato_WGCNA_modules_mean_z_reordered <- tomato_WGCNA_modules_mean_z %>% 
  inner_join(Module_eigen_long_peak %>% 
               mutate(module = str_remove(Module, "ME")), 
             by = "module") %>% 
  mutate(module = reorder(module, -order_y))

head(tomato_WGCNA_modules_mean_z_reordered)
```

```{r}
tomato_WGCNA_modules_mean_z_reordered$mean.z %>% summary()
quantile(tomato_WGCNA_modules_mean_z_reordered$mean.z, c(0.025, 0.975))
```


```{r}
tomato_WGCNA_heatmap.z <- tomato_WGCNA_modules_mean_z_reordered %>% 
  mutate(mean.z.clipped = case_when(
    mean.z > 1.5 ~ 1.5,
    mean.z < -1.5 ~ -1.5,
    T ~ mean.z
  )) %>% 
  ggplot(aes(x = tissue, y = module)) +
  facet_grid(.~ dev_stage, scales = "free", space = "free") +
  geom_tile(aes(fill = mean.z.clipped), color = "grey80") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdBu")), limits = c(-1.5, 1.5),
                       breaks = c(-1.5, 0, 1.5), labels = c("< -1.5", "0", "> 1.5")) +
  labs(x = NULL,
       y = "Module",
       fill = "z score") +
  theme_classic() +
  theme(
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    axis.text.x = element_blank(),
    strip.text = element_blank(),
    legend.position = "top",
    panel.spacing = unit(0.5, "lines") 
  )

wrap_plots(tomato_WGCNA_heatmap.z, heat_map_color_strip_y,
           heatmap_color_strip1, blank,
           heatmap_color_strip2, blank, 
           nrow = 3, ncol = 2, 
           heights = c(1, 0.05, 0.05),
           widths = c(1, 0.02), guides = "collect") &
  theme(
    legend.position = "bottom",
    legend.box = "vertical"
  )

ggsave("../Results/WGCNA_tomato_heatmap.z.svg", height = 4.8, width = 10, bg = "white")
ggsave("../Results/WGCNA_tomato_heatmap.z.png", height = 4.8, width = 10, bg = "white")
```
# Cross-workflow comparison
## Correspondance

### Wide of WGCNA
```{r}
tomato_WGCNA_modules_mean_z_reordered_wide <- tomato_WGCNA_modules_mean_z_reordered %>% 
  select(dev_stage, tissue, module, mean.z) %>% 
  mutate(tag = paste(tissue, dev_stage, sep = ";")) %>% 
  select(-tissue, -dev_stage) %>% 
  pivot_wider(names_from = module, values_from = mean.z) %>% 
  as.data.frame()

row.names(tomato_WGCNA_modules_mean_z_reordered_wide) <- tomato_WGCNA_modules_mean_z_reordered_wide$tag
head(tomato_WGCNA_modules_mean_z_reordered_wide)
```

### Wide of Simple Tidy GeneCoEx
```{r}
modules_mean_z_reorded_wide <- modules_mean_z_reorded %>% 
  select(dev_stage, tissue, module, mean.z) %>% 
  mutate(tag = paste(tissue, dev_stage, sep = ";")) %>% 
  select(-tissue, -dev_stage) %>% 
  pivot_wider(names_from = module, values_from = mean.z) %>% 
  as.data.frame()

row.names(modules_mean_z_reorded_wide) <- modules_mean_z_reorded_wide$tag
head(modules_mean_z_reorded_wide)
```

```{r}
modules_mean_z_reorded_wide$tag == tomato_WGCNA_modules_mean_z_reordered_wide$tag
```
### Find corresponding modules
```{r}
Corresonding_modules <- apply(tomato_WGCNA_modules_mean_z_reordered_wide[,-1], 
                              FUN = stats::cor, modules_mean_z_reorded_wide[,-1], 
                              MARGIN = 2)

dim(Corresonding_modules)
head(Corresonding_modules)
```

```{r}
Corresonding_modules_long <- Corresonding_modules %>% 
  as.data.frame() %>% 
  cbind(SimpleTidy_GeneCoEx = colnames(modules_mean_z_reorded_wide)[-1]) %>% 
  pivot_longer(cols = ! SimpleTidy_GeneCoEx, names_to = "WGCNA", values_to = "r") %>% 
  inner_join(module_peak_exp %>% 
               mutate(module = as.character(module)) %>% 
               select(dev_stage, tissue, module, order_y, peak_exp), 
             by = c("SimpleTidy_GeneCoEx" = "module")) %>% 
  rename(order_simple = order_y) %>% 
  inner_join(
    tomato_WGCNA_modules_mean_z_reordered %>% 
  group_by(module, order_y, peak_stage) %>% 
  count() %>% 
    select(-n), 
  by = c("WGCNA" = "module")
  ) %>% 
  rename(order_WGCNA = order_y) %>% 
  mutate(SimpleTidy_GeneCoEx = reorder(SimpleTidy_GeneCoEx, -order_simple)) %>% 
  mutate(WGCNA = reorder(WGCNA,  order_WGCNA))


head(Corresonding_modules_long)
```

```{r}
Corresonding_modules_long %>% 
  filter(WGCNA == "blue") %>% 
  filter(SimpleTidy_GeneCoEx == "5")

Corresonding_modules_long %>% 
  filter(WGCNA == "brown") %>% 
  filter(SimpleTidy_GeneCoEx == "9")
```


```{r}
Corresonding_modules_heatmap <- Corresonding_modules_long %>% 
  ggplot(aes(x = SimpleTidy_GeneCoEx, y = WGCNA)) +
  geom_tile(aes(fill = r)) +
  scale_fill_gradientn(colours = rev(brewer.pal(11, "RdBu")),
                       limits = c(-1, 1),
                       breaks = c(-0.9, 0, 0.9))+
  labs(x = "Simple Tidy GeneCoEx Modules",
       y = "WGCNA Modules") +
  theme_minimal() +
  theme(
    legend.position = "top",
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    #axis.text.y = element_blank(),
    legend.key.width = unit(1.2, "lines"),
    panel.grid = element_blank()
  )


Corresonding_modules_heatmap

ggsave("../Results/Tomato_correspondance.svg", height = 5, width = 7)
ggsave("../Results/Tomato_correspondance.png", height = 5, width = 7)
```



## Squared error loss 
### Simple Tidy Gene CoEx
```{r}
Tomato_tidy_msqs <- Exp_table_long_averaged_z_high_var_modules %>% 
  group_by(module, tissue, dev_stage) %>% 
  mutate(mean = mean(z.score)) %>% 
  mutate(squares = (z.score - mean)^2) %>% 
  ungroup() %>% 
  group_by(module) %>% 
  summarise(
    ssq = sum(squares)
  ) %>% 
  ungroup() %>% 
  inner_join(
    my_network_modules %>% 
      group_by(module) %>% 
      count(),
    by = "module"
  ) %>% 
  mutate(msq = ssq/n)

Tomato_tidy_msqs
```

### WGCNA
```{r}
Tomato_WGCNA_msq <-Exp_table_long_averaged_modules %>% 
 # filter(module != "grey") %>% 
  group_by(module, tissue, dev_stage) %>% 
  mutate(mean = mean(z.score)) %>% 
  mutate(squares = (z.score - mean)^2) %>% 
  ungroup() %>% 
  group_by(module) %>% 
  summarise(
    ssq = sum(squares)
  ) %>% 
  ungroup() %>% 
  inner_join(
    module_df %>% 
      group_by(module) %>% 
      count(),
    by = "module"
  ) %>% 
  mutate(msq = ssq/n) %>% 
  filter(is.na(msq) == F)

head(Tomato_WGCNA_msq)
```

### Comparison
```{r}
lost_function_comparisons <- rbind(
  Tomato_tidy_msqs %>% 
    select(msq, n) %>% 
    mutate(method = "Simple Tidy"),
  Tomato_WGCNA_msq %>% 
    select(msq, n) %>% 
    mutate(method = "WGCNA")
)

lost_function_comparisons_s <-lost_function_comparisons %>% 
  group_by(method) %>% 
  summarise(
    mean = mean(msq),
    median = median(msq),
    sd = sd(msq),
    NN = n()
  )

lost_function_comparisons_s
```

```{r}
wilcox.test(lost_function_comparisons$msq ~ lost_function_comparisons$method)
cor.test(lost_function_comparisons$msq, lost_function_comparisons$n)
```



```{r}
median_separation <- lost_function_comparisons %>% 
  ggplot(aes(x = method, y = msq)) +
  # geom_boxplot(width = 0.3) +
  ggbeeswarm::geom_quasirandom(aes(fill = method), size = 3,
                                shape = 21, alpha = 0.8, color = "white") +
  scale_fill_manual(values = c("tomato1", "grey30")) +
  labs(x = "method",
       y = "loss function\n(mean sum of squares)",
       title = "Data from Shinozaki et al., 2018",
       caption = paste0(
         "median1 = ", signif(lost_function_comparisons_s[1, 3], 3), "; ",
         "median2 = ", signif(lost_function_comparisons_s[2, 3], 3), "\n",
         "P = ", 
         signif(
           wilcox.test(lost_function_comparisons$msq ~ lost_function_comparisons$method)$p.value,
           2), 
         "\n(Wilcoxon Rank Sum Test)"
       )) +
  theme_classic()+
  theme(
    legend.position = "none",
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    plot.title = element_text(size = 12),
    plot.caption = element_text(size = 12, hjust = 0)
  )

msq_n_scatter <- lost_function_comparisons %>% 
  ggplot(aes(x = n,  y = msq)) +
  geom_point(aes(fill = method), size = 3,
                                shape = 21, alpha = 0.8, color = "white") +
  scale_fill_manual(values = c("tomato1", "grey30")) +
  labs(
    y = "loss function\n(mean sum of squares)",
    x = "Num. genes in module",
    caption = paste0(
      "r = ", signif(
        cor.test(lost_function_comparisons$msq, lost_function_comparisons$n)$estimate, 3
      )
    )
  ) +
  theme_classic() +
  theme(
    legend.position = c(0.8, 0.2),
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    plot.title = element_text(size = 12),
    plot.caption = element_text(size = 12, hjust = 0)
  )

wrap_plots(median_separation,
           msq_n_scatter,
           nrow = 1)

ggsave("../Results/Tomato_benchmarking_results.svg", height = 4, width = 7, bg = "white")
ggsave("../Results/Tomato_benchmarking_results.png", height = 4, width = 7, bg = "white")
```
## Controlling for module size
```{r}
model_tomato <- lm(msq ~ n + method, data = lost_function_comparisons)
anova(model_tomato)
```



# Membership comparison
```{r}
module_df
my_network_modules
```

```{r}
find_overlap <- function(SimpleTidy_module, WGCNA_module){
   x = my_network_modules %>% 
     filter(module == SimpleTidy_module)
   
   y = module_df %>% 
     rename(gene_ID = gene.id) %>% 
     filter(module == WGCNA_module)

   intersect(x$gene_ID, y$gene_ID) %>% 
     length()
}

find_overlap(
  SimpleTidy_module = "5",
  WGCNA_module = "blue"
)
```

```{r}
module_combinations <- expand.grid(
   SimpleTidy_module = unique(my_network_modules$module),
   WGCNA_module = unique(module_df$module)
)

membership_overlap <- mapply(FUN = find_overlap, 
       module_combinations$SimpleTidy_module, 
       module_combinations$WGCNA_module) %>% 
  as.data.frame() %>% 
  rename(overlap = ".") %>% 
  cbind(module_combinations)

head(membership_overlap)
```


```{r}
membership_overlap_annotated <- membership_overlap %>% 
  mutate(Module = paste0("ME", WGCNA_module)) %>% 
  inner_join(module_df_colors, by = "Module") %>% 
  inner_join(module_peak_exp, by = c("SimpleTidy_module" = "module"))

head(membership_overlap_annotated)
```


## Alluvial plot

```{r}
membership_overlap_annotated %>% 
  mutate(peak_exp = as.character(peak_exp)) %>% 
  mutate(stage = case_when(
    str_detect(peak_exp, "MG|Br|Pk") ~ str_sub(peak_exp, start = 1, end = 2),
    T ~ peak_exp
  )) %>% 
  mutate(stage = factor(stage, levels = c(
   "5 DPA",
   "10 DPA",
   "20 DPA",
   "30 DPA",
   "MG",
   "Br",
   "Pk",
   "LR",
   "RR"
  ))) %>% 
  ggplot(aes(y = overlap, 
           axis1 = as.factor(SimpleTidy_module), 
           axis2 = colors)) +
  geom_alluvium(aes(fill = stage), alpha = 0.8) +
  geom_stratum(fill = "grey70", color = "grey30", width = 0.05) +
  scale_x_discrete(labels = c("Simple Tidy", "WGCNA"),
                   limits = c("Simple Tidy", "WGCNA"),
                   expand = c(0.01, 0.01)) +
  scale_fill_manual(values = viridis(9, option = "D")[c(-4, -8)]) +
  labs(x = "method",
       y = "number of genes",
       fill = "peak\nexpression") +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    text = element_text(size = 14),
    axis.text = element_text(color = "black"),
    panel.grid = element_blank()
  ) +
  coord_flip()

ggsave("../Results/WGCNA_tidy_memebership.svg", height = 3, width = 8, bg = "white")
ggsave("../Results/WGCNA_tidy_memebership.png", height = 3, width = 8, bg = "white")
```

